File: starter_templates/sqlite/python/app/record_parser.py
from .varint_parser import parse_varint


def parse_record(stream, column_count):
    """
    Parses SQLite's "Record Format" as mentioned here: https://www.sqlite.org/fileformat.html#record_format
    """
    _number_of_bytes_in_header = parse_varint(stream)

    serial_types = [parse_varint(stream) for i in range(column_count)]
    return [parse_column_value(stream, serial_type) for serial_type in serial_types]


def parse_column_value(stream, serial_type):
    if (serial_type >= 13) and (serial_type % 2 == 1):
        # Text encoding
        n_bytes = (serial_type - 13) // 2
        return stream.read(n_bytes)
    elif serial_type == 1:
        # 8 bit twos-complement integer
        return int.from_bytes(stream.read(1), "big")
    else:
        # There are more cases to handle, fill this in as you encounter them.
        raise Exception(f"Unhandled serial_type {serial_type}")

File: starter_templates/sqlite/go/app/record_parser.go
package main

import (
	"io"
	"log"
)

type Record struct {
	values []interface{}
}

// parseRecord parses SQLite's "Record Format", as mentioned here: https://www.sqlite.org/fileformat.html#record_format
func parseRecord(stream io.Reader, valuesCount int) Record {
	parseVarint(stream) // number of bytes in header

	serialTypes := make([]int, valuesCount)

	for i := 0; i < valuesCount; i++ {
		serialTypes[i] = parseVarint(stream)
	}

	values := make([]interface{}, valuesCount)

	for i, serialType := range serialTypes {
		values[i] = parseRecordValue(stream, serialType)
	}

	return Record{values: values}
}

func parseRecordValue(stream io.Reader, serialType int) interface{} {
	if serialType >= 13 && serialType%2 == 1 {
		// Text encoding
		bytesCount := (serialType - 13) / 2
		value := make([]byte, bytesCount)
		_, _ = stream.Read(value)
		return value
	} else if serialType == 1 {
		// 8 bit twos-complement integer
		return parseUInt8(stream)
	} else {
		// There are more cases to handle, fill this in as you encounter them.
		log.Fatalf("Unknown serial type %v", serialType)
		return nil
	}
}
